plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/analysers/DeadCodeFinder.java

```diff
@@ -0,0 +1,280 @@
+		if (preds.isEmpty()) {
+			return true;
+		}
+
+		Set<ControlFlowElement> visited = new HashSet<>();
+		while (!preds.isEmpty()) {
+			ControlFlowElement pred = preds.iterator().next();
+			preds.remove(pred);
+			if (visited.contains(pred))
+				continue;
+
+			if (allForwardCFEs.contains(pred)) {
+				return false;
+			}
+			preds.addAll(flowAnalyses.getPredecessorsSkipInternal(pred));
+			visited.add(pred);
+		}
+		return true;
+	}
+
+	/** @return all {@link TextRegion}s of dead code */
+	public Set<DeadCodeRegion> getDeadCodeRegions() {
+		Collection<Set<ControlFlowElement>> deadCodeGroups = separateOnTheirBlocks(unreachableCFEs);
+		Set<DeadCodeRegion> deadCodeRegions = new HashSet<>();
+		for (Set<ControlFlowElement> deadCodeGroup : deadCodeGroups) {
+			DeadCodeRegion textRegion = getDeadCodeRegion(deadCodeGroup);
+			deadCodeRegions.add(textRegion);
+		}
+		return deadCodeRegions;
+	}
+
+	/**
+	 * Separates the given set into sets where all {@link ControlFlowElement}s of each set have the same containing
+	 * {@link Block}.
+	 * <p>
+	 * Note that the assumption is:<br/>
+	 * <i>No block can contain more than one single dead code region.</i>
+	 */
+	private Collection<Set<ControlFlowElement>> separateOnTheirBlocks(Set<ControlFlowElement> unreachableElems) {
+		Map<Block, Set<ControlFlowElement>> unreachablesMap = new HashMap<>();
+		for (ControlFlowElement unreachableElem : unreachableElems) {
+			HashSet<ControlFlowElement> moreUnreachableElems = new HashSet<>();
+			Block cfeBlock = getReachableBlock(unreachableElems, unreachableElem, moreUnreachableElems);
+			if (cfeBlock == null)
+				continue;
+
+			if (!unreachablesMap.containsKey(cfeBlock)) {
+				unreachablesMap.put(cfeBlock, new HashSet<>());
+			}
+			Set<ControlFlowElement> unreachableInBlock = unreachablesMap.get(cfeBlock);
+			unreachableInBlock.add(unreachableElem);
+			unreachableInBlock.addAll(moreUnreachableElems);
+		}
+
+		return unreachablesMap.values();
+	}
+
+	/** Finds the nearest reachable {@link Block} of the given {@link ControlFlowElement} */
+	private Block getReachableBlock(Set<ControlFlowElement> unreachableElems, ControlFlowElement unreachableElem,
+			Set<ControlFlowElement> moreUnreachableElems) {
+
+		EObject elemContainer = unreachableElem.eContainer();
+		if (elemContainer instanceof ExpressionStatement) {
+			moreUnreachableElems.add((ExpressionStatement) elemContainer);
+		}
+
+		Block block = EcoreUtil2.getContainerOfType(unreachableElem, Block.class);
+		if (block == null) // can be null in case of broken ASTs
+			return null;
+
+		EObject blockContainer = block.eContainer();
+		boolean isDeadContainer = blockContainer instanceof ControlFlowElement;
+		isDeadContainer &= isDeadContainer && FGUtils.isControlElement((ControlFlowElement) blockContainer);
+		isDeadContainer &= isDeadContainer && isDeadCode((ControlFlowElement) blockContainer);
+
+		if (isDeadContainer) {
```

Comment:
a) I prefer this style because I want to avoid nested if blocks. Also, I prefer this, because the condition has a distinct name and is - in my opinion - far better understandable than some succeeding and nested if conditions.

b) A _Control Element_ is a statement like `for`, `if` or `try` that has no side effect on its own. Maybe I rename it to `isControlStatement`.